text/sprop.md

```diff
@@ -0,0 +1,500 @@
+landing in irrelevant types.
+
+As a workaround, I allow cumulativity during elaboration. At
+definition time the kernel will check that filling in the evars
+produced a non-cumulativity-using value.
+
+However this opens us to issues with the marks: at the time the binder
+for `x` is generated and when we add `x` to the environment in `fun x
+=> (x : A)` we are treating `x` as a relevant variable. The
+consequences are
+
+- the terms generated are incorrect and the kernel will error on them.
+  Since the kernel has the information we deal with this issue by
+  making it repair the terms.
+
+- some irrelevance may be missed at unification time as some marks are
+  wrongly `Relevant`. eg in `fun (A:SProp) (P:A->SProp) x (v:P x) y =>
+  v : P y` when checking `v : P y` the types for `x` and `y` have been
+  unified to `A : SProp` but their marks are still incorrect.
+
+  Currently the only solution is for the user to workaround by
+  annotating more.
+
+# Inductive types
+
+In current Coq inductive types live in the max of the universes of
+their constructor arguments (or larger with cumulativity). For `Prop`
+types they must also satisfy singleton elimination, i.e. they must
+have 0 or 1 constructor.
+
+An inductive type can be defined without satisfying this rule in an
+impredicative universe, but it may not be eliminated to types living
+in other universes (with cumulativity this means types living in
+higher universes).
+
+When adding `SProp` (which is impredicative) these rules still hold
+with the following modifications:
+
+- non-`SProp` inductive types may have constructor arguments in
+  `SProp`. This allows e.g.
+  ~~~coq
+  Inductive Box (A:SProp) : Prop := box : A -> Box A.
+  ~~~
+
+  The inductive type stops definitional proof irrelevance from
+  escaping `SProp`: we don't have `forall x y : Box A, x = y`
+  definitionally (we do have it propositionally).
+
+- as an exception to the above, primitive records whose fields are all
+  `SProp` must live in `SProp` (we downgrade them to emulated records
+  otherwise, like with records with non-definable projections).
+  Otherwise proof irrelevance would escape through η.
+
+- primitive records may live in `SProp` when all their fields are
+  `SProp` (otherwise they get squashed, then projections are
+  non-definable so they get downgraded to emulated records)
+
+- regular inductive types may live unsquashed in `SProp` when they
+  satisfy some condition. See next section.
+
+# Conditions for unsquashed inductive types living in `SProp`
+
+## Basic types
+
+Here is the most important non-squashed `SProp` type:
+~~~coq
+Inductive sEmpty : SProp := .
+~~~
+Without it we can't even be sure that `SProp` has non trivial types!
+
+Next are some types which we have to squash:
+~~~coq
+Inductive sbool : SProp := strue | sfalse.
+(* inconsistent theory when unsquashed *)
+
+Inductive Pointed (A:Type) (a:A) : SProp := { point : A; eq : point = x; }.
```

Comment:
there are some typos here,eg x vs a.
also the first equation seems to compare terms of different types (pointed x) vs  (pointed y)...

Reply:
could you also write down the types of the two terms being compared, or point to the relevant section of your paper? Eg, what do you mean by irrelevance (of a `p : x =_A y` for any `A`)?